### DISCUSSION TITLE
Fetch api only work on simulator

### DISCUSSION THREAD

**Discussion by Rafucho25** [zepp-health/zeppos-forum#108] [Developer Feedback]
Hi, my problem is when i call an api in the simulator it works great but in the device always return undefined. Here are some pictures:
![band_emulator](https://user-images.githubusercontent.com/16562078/224373343-072e3dc1-cf08-4ec8-8e37-f0665e70f766.png)
Band 7 in simulator

![log_simulator](https://user-images.githubusercontent.com/16562078/224373423-531f3163-aeb9-4a88-bf79-3a5cc4e2b426.png)
Log in simulator

![band_device](https://user-images.githubusercontent.com/16562078/224373493-72a6e78e-f780-4943-9ba9-c1cc59b0c28b.jpg)
My band 7  

![log_device](https://user-images.githubusercontent.com/16562078/224373591-22f75fdc-e3a5-45ad-b73c-a405f10084d2.jpg)
Log in my smartphone

This is the code:

index.js in app-side:

```javascript
const fetchData = async (ctx, _url, _header) => {
  try {
    // Requesting network data using the fetch API
    const res = await fetch({
      url: _url,
      method: 'GET',
      headers: _header
    })
    
    ctx.response({
      data: { result: JSON.parse(res.body)  },
    })
    console.log(data)

  } catch (error) {
    ctx.response({
      data: { result: 'ERROR' },
    })
  }
}
```

index.js in page:

```javascript
  build() {
    messageBuilder.request({
      method: "GET_DATA",
      url: `https://api.exchangerate.host/convert?from=USD&to=DOP&amount=100`,
      header: {
        "Accept": "*/*"
      }
    })
    .then(data => {
      const { result = {} } = data
      
      console.log(result.result);
      hmUI.createWidget(hmUI.widget.TEXT, {
        y: 93,
        w: 184,
        color: 0xffffff,
        text_size: 24,
        align_h: hmUI.align.CENTER_H,
        text_style: hmUI.text_style.NONE,
        text: result.result
      })
      
    })
  }
```
Result of the api:

```javascript
{
  "motd": {
    "msg": "If you or your company use this project or like what we doing, please consider backing us so we can continue maintaining and evolving this project.",
    "url": "https://exchangerate.host/#/donate"
  },
  "success": true,
  "query": {
    "from": "USD",
    "to": "DOP",
    "amount": 100
  },
  "info": {
    "rate": 55.082327
  },
  "historical": false,
  "date": "2023-03-10",
  "result": 5508.2327
}
```

My smartphone run android 10 and device is band 7 system version: 1.10.58.1

**Comment by alexmathew11** [comment]
I think official zepp developers won't gonna answer this because mi band 7 is not an official zepp os device.

**Reply by Rafucho25** [reply]
The device is amazfit band 7, sorry if i wasnt clear.

**Comment by alexmathew11** [comment]
I can't understand the problem but if official developers helped you so and if it would work flawlessly so give me the correct code of I also wanted to try

**Comment by MyoungXUE** [comment]
Hi, sorry for the long wait, this may be a bug in the fetch API.

`res.body` may be a JS Object instead of a string in some devices, in which case `JSON.stringify` is not needed, so it is recommended to add a type determination.

For example:

```js
ctx.response({
  data: { result:  typeof res.body === 'string' ?  JSON.parse(res.body) : res.body  },
})
```

**Reply by Rafucho25** [reply]
Hi, this works perfect. Thanks for the help to all of you.

**Reply by bsavarin** [reply]
@MyoungXUE Thanks for this - I was also asking and this solution has been most helpful.  Can the code be corrected in the sample to help others please?

**Reply by MyoungXUE** [reply]
> @MyoungXUE Thanks for this - I was also asking and this solution has been most helpful. Can the code be corrected in the sample to help others please?

Hi, this is a very good suggestion, we will update the sample code and update the documentation this week.
